proposals/function-pointers.md

```diff
@@ -0,0 +1,245 @@
+# Function Pointers
+
+## Summary
+This proposal privodes language constructs that expose IL opcodes that cannot currently be accessed efficiently,
+or at all, in C# today: `ldftn` and `calli`. These IL opcodes can be important in high performance code and developers
+need an effecient way to access them.
+
+## Motivation
+The motivations and background for this feature are described in the following issue (as is a 
+potential implementation of the feature): 
+
+https://github.com/dotnet/csharplang/issues/191
+
+This is an alternate design propsoal to [compiler intrinsics](https://github.com/dotnet/csharplang/blob/master/proposals/intrinsics.md)
+
+## Detailed Design 
+
+### funcptr
+The language will allow for the declaration of function pointers using the `funcptr` contextual keyword. The 
+declaration and usage of function pointers closely resemble that of `delegate`:
+
+``` csharp
+unsafe class Example {
+    delegate void DAction(int a);
+    funcptr void FAction(int a);
+
+    void Example(DAction d, FAction f) {
+        d(42);
+        f(42);
+    }
+}
+```
+
+These types are represented using the function pointer type as outlined in ECMA-335. This means invocation
+of a `funcptr` will use `calli` where invocation of a `delegate` will use `callvirt` on the `Invoke` method.
+Syntactically though invocation is identical for both constructs.
+
+The `calli` instruction requires the calling convention be specified as a part of the invocation. The default 
+for `funcptr` will be managed. Alternate forms can be specified by adding the appropriate modifier after the 
+`funcptr` keyword: `cdecl`, `fastcall`, `stdcall`, `thiscall` or `winapi`. Example:
```

Comment:
We should have identifier for the default unmanaged calling convention - what you get using `default(CallingConvention)` today. Would `funcptr unmanaged` work?

winapi maps to default unmanaged calling convention, but it is not a very good name because if it is Windows-specific. It should be fine to not have it.

Reply:
Overall that seems fine. The proposal today though doesn't rely on the type `CallingConvention`. Is there a different way to calculate the default?

Reply:
Hmmm, the enum that describes calling conventions in signatures does not have neither `winapi` nor good value for default unmanaged calling convention:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/SignatureCallingConvention.cs#L22

Maybe we should consider adding value for the default unmanaged calling convention here? (Of course, it would work on new runtimes only...)